foxai/explainer/fullgradeints.py

```diff
@@ -0,0 +1,462 @@
+
+        return input_gradients, self.feature_grads
+
+
+class FullGrad:
+    """
+    Compute FullGrad saliency map and full gradient decomposition
+    """
+
+    def __init__(self, model, image_size=(3, 224, 224)):
+        self.model: torch.nn.Module = model
+        self.im_size: Tuple[int, int, int, int] = (1,) + image_size
+        self.model_ext: FullGradExtractor = FullGradExtractor(
+            model=self.model, im_size=image_size
+        )
+        self._biases: List[torch.Tensor] = self.model_ext.biases
+
+        cuda = next(self.model.parameters()).is_cuda
+        self._device = torch.device("cuda" if cuda else "cpu")
+        _ = self.check_completeness()
+
+    def check_completeness(self) -> bool:
+        """Check if completeness property is satisfied. If not, it usually means that
+        some bias gradients are not computed (e.g.: implicit biases of non-linearities).
+
+        Returns:
+            whether the completnes condition is fulfielled
+        """
+
+        # Random input image
+        sample_input: torch.Tensor = torch.randn(self.im_size).to(self._device)
+
+        # Get raw outputs
+        self.model.eval()
+        try:
+            raw_output = self.model(sample_input)
+        except RuntimeError as e:
+            log().error(e)
+            log().error(_INPLACE_ERROR_MSG)
+            terminate()
+
+        # Compute full-gradients and add them up
+        # if class is None, the argmax of output is chosen
+        input_grad, bias_grad = self.full_gradient_decompose(
+            image=sample_input, target_class=None
+        )
+
+        fullgradient_sum = (input_grad * sample_input).sum()
+        for i in range(len(bias_grad)):
+            fullgradient_sum += bias_grad[i].sum()
+
+        # Compare raw output and full gradient sum
+        if not isclose(raw_output.max().item(), fullgradient_sum.item(), rel_tol=1e-2):
+            log().warning(
+                "Completness not fullfield. Incorrect computation of bias-gradients, "
+                "most probably due to using activation functions other than ReLU. The XAI map may be incorect."
+                f"Raw output = {str(raw_output.max().item())} Full-gradient sum = {str(fullgradient_sum.item())}"
+            )
+
+            return False
+
+        return True
+
+    def full_gradient_decompose(
+        self, image: torch.Tensor, target_class: Optional[torch.Tensor] = None
+    ) -> Tuple[torch.Tensor, List[torch.Tensor]]:
+        """
+        Compute full-gradient decomposition for an input image.
+
+        Args:
+            image: the input image to the nn
+            target_class: the class for which to draw explanations. If None, the argmax of output is chosen.
+
+        Returns:
+            input image gradients (contribution of each input pixel to the final output)
+            list of contributions of bias terms in each layer in the model (each layer with the bias term)
+        """
+
+        self.model.eval()
```

Comment:
Same as above, we should somehow revert this change to the model.

Reply:
done